app/apps/ood_app.rb

```diff
@@ -40,6 +42,29 @@ def url
     end
   end
 
+  # the problem is we need the context :-P
+  def links
+    if role == "files"
+      # assumes Home Directory is primary...
+      [Link.new("Home Directory", OodAppkit.files.url(path: Dir.home), 'home')] + OodFilesApp.new.favorite_paths.map do |path|
+        Link.new(path.to_s, OodAppkit.files.url(path: path), "folder")
+      end
+    elsif role == "shell"
+      if ApplicationController.helpers.login_clusters.count == 0
+        [Link.new("Shell Access", OodAppkit.shell.url,"terminal")]
+      else
+        ApplicationController.helpers.login_clusters.map { |c| Link.new("#{c.metadata.title} Shell Access", OodAppkit.shell.url(host: c.login.host), "terminal") }
```

Comment:
This should probably be:

```ruby
ApplicationController.helpers.login_clusters.map do |c|
  Link.new("#{c.metadata.title || c.id.to_s.titleize} Shell Access", OodAppkit.shell.url(host: c.login.host), "terminal")
end
```

as `Cluster#metadata` has attributes that could be `nil`.